### ISSUE TITLE
Onesignal and Firebase Database don't work together

### ISSUE THREAD

**Issue by pranavt7** [OneSignal/OneSignal-Unity-SDK#119]
I need Firebase Database and OneSignal to work together for my app. Whenever I import any one of them, Unity Prompts me the following -

The following obsolete files are referenced by packages in your project, would you like to delete them?
```
Assets/Firebase/Editor/FirebaseAuth_v4.5.2_manifest.txt: Assets/PlayServicesResolver/Editor/Google.JarResolver_v1.2.64.0.dll
Assets/Firebase/Editor/FirebaseDatabase_v4.5.2_manifest.txt: Assets/PlayServicesResolver/Editor/Google.JarResolver_v1.2.64.0.dll
Assets/Firebase/Editor/FirebaseAuth_v4.5.2_manifest.txt: Assets/PlayServicesResolver/Editor/Google.VersionHandlerImpl_v1.2.64.0.dll
Assets/Firebase/Editor/FirebaseDatabase_v4.5.2_manifest.txt: Assets/PlayServicesResolver/Editor/Google.VersionHandlerImpl_v1.2.64.0.dll
Assets/Firebase/Editor/FirebaseAuth_v4.5.2_manifest.txt: Assets/PlayServicesResolver/Editor/Google.IOSResolver_v1.2.64.0.dll
Assets/Firebase/Editor/FirebaseDatabase_v4.5.2_manifest.txt: Assets/PlayServicesResolver/Editor/Google.IOSResolver_v1.2.64.0.dll
```

Firebase works on the UnityEditor. But when I make a build on Android, both Firebase and One Signal don't work. 

Any ideas as to what might the issue be?
Thanks!

**Work Around:** https://github.com/OneSignal/OneSignal-Unity-SDK/issues/119#issuecomment-415283191

**Comment by pranavt7** [comment]
But, the problem is there aren't any build errors. I am able to make a build, but it just doesn't seem to work in a device. Firebase and OneSignal aren't initialising on a device.(No Output)

**Comment by Nightsd01** [comment]
I don't really understand what you mean when you say "it doesn't work". 

When you say OneSignal specifically doesn't work, what happens? Can you please post a link to a full logcat from at least the first 10 seconds of your app running? Even without errors we can usually see something. 

Please make sure to set the OneSignal log level to `verbose`

**Comment by Argurth** [comment]
We spotted a similar problem after updating our SDK to manage GDPR consent. Before the update, everithing is fine. After updating only the OneSignal SDK (from 2.6.4 to 2.7.2), everything is still working fine in the editor, but Firebase won't init anymore, without throwing any exception. Maybe it's related to "Now supports FCM (Firebase Cloud Messaging) library"

**Comment by Argurth** [comment]
We spotted a similar problem after updating our SDK to manage GDPR consent.

Before the update, everything is fine. After updating only the OneSignal SDK (from 2.6.4 to 2.7.2), everything is still working fine in the editor, but Firebase won't init anymore, without throwing any exception. Maybe it's related to "Now supports FCM (Firebase Cloud Messaging) library"

**Comment by Argurth** [comment]
We spotted a similar problem after updating our SDK to manage GDPR consent. Before the update, everything was working fine. 

After updating only the OneSignal SDK (from 2.6.4 to 2.7.2), everything is still working fine in the editor, but Firebase won't init anymore on Android, without throwing any exception. 

Maybe it's related to "Now supports FCM (Firebase Cloud Messaging) library" of the 2.7.0 version., because we noticed that removing the following depency resolved the problem:

`<androidPackage spec="com.google.firebase:firebase-messaging:[10.2.1, 12.1.0)"/>`

The Firebase version we're using is 4.5.2

Hope it will help to resolve this issue !

**Comment by Russel1** [comment]
similar problem here:
https://github.com/OneSignal/OneSignal-Unity-SDK/issues/123

**Comment by Argurth** [comment]
Removing the depency isn't enough, however, it works if it is remplaced with this one : 

<androidPackage spec="com.google.android.gms:play-services-gcm:[10.2.1, 12.1.0)" />

I think it should work as a temporary fix until OneSignal fix the problem

**Comment by Russel1** [comment]
it still occurs in 2.7.3.

**Comment by Saoish** [comment]
I have encountered the same problem. Furthermore, I found that Google Sign-in and Twitter Login SDK no longer work as long as I have OneSignal SDK integrated into the project.

**Comment by jkasten2** [comment]
All the files under `Assets/PlayServicesResolver/` should have version 1.2.75.0 or higher such as `Google.JarResolver_v1.2.75.0.dll`. There was a recent update that fixes issues with the wrong .aar files being downloaded in the `Assets/Plugins/Android/` folder when there are a number of plugins requiring versions above and below version 15 of Firebase and GMS. The OneSignal 2.7.4 SDK includes this version we recommend reimporting OneSignal as another plugin could have downgraded the PlayServicesResolver.

Please make sure to rerun Assets > Play Services Resolver > Android Resolver > Force Resolve after reimporting. If you are still having an issue then provide the following.
1. A list of plugins and version you are using in your project
2. A list of all files in your `Assets/Plugins/Android/` folder
3. The exact build or runtime error you are seeing

**Comment by yo0n** [comment]
Hi, we're having similar issues with OneSignal SDK. With OneSignal integrated, it seems that Firebase don't initialize. We've tested before integrating the SDK, after integrating the SDK and after removing the SDK.

When integrating OneSignal, we followed the Unity Setup Guide from OneSignal Docs and the reimport checklist that @jkasten2 mentioned it earlier. We made sure that we run Assets>PlayServicesResolver>Android Resolver>Force Resolve.

> If you are still having an issue then provide the following.
> 
> 1. A list of plugins and version you are using in your project
> 2. A list of all files in your Assets/Plugins/Android/ folder
> 3. The exact build or runtime error you are seeing

As we are still having issues with it:
1. -[play-games-plugins-for-unity 0.9.50](https://github.com/playgameservices/play-games-plugin-for-unity)
    -[IronSource Ad Mediation SDK](https://www.ironsrc.com/)
    -[Firebase Analytics 4.5.0](https://firebase.google.com/docs/analytics/unity/start)
    -[Firebase RemoteConfig 4.5.0](https://firebase.google.com/docs/remote-config/use-config-unity)
2. ![list of all files in Assets/Plugins/Android (1/3)](https://user-images.githubusercontent.com/22488012/42601036-ace73f96-853a-11e8-9ccb-9468d6311558.png)
![list of all files in Assets/Plugins/Android (2/3)](https://user-images.githubusercontent.com/22488012/42601039-b008ce92-853a-11e8-91da-af4e418b7f37.png)
![list of all files in Assets/Plugins/Android (3/3)](https://user-images.githubusercontent.com/22488012/42601040-b1c207a8-853a-11e8-9e1d-f7f1276aec49.png)
3. As @Saoish mentioned, "There is no build or run time error. On android device, Firebase could not been initialized." this is what seems to be happening with us too.

And as instructed, we are using player-service-resolver v1.2.75.0.
![screenshot_81](https://user-images.githubusercontent.com/22488012/42601219-46a62e08-853b-11e8-82ce-578e70cc01c8.png)

Here is a [log.txt](https://github.com/OneSignal/OneSignal-Unity-SDK/files/2186317/log.txt) from our device.

Some more infos:
Unity version: 5.6.5p2
Device: [Samsung Galaxy S III](https://www.gsmarena.com/samsung_i9300_galaxy_s_iii-4238.php)
Java JDK: 1.8.0_161
NDK: r13b

**Comment by Russel1** [comment]
I have the same issue "RegistrationIntentService not found"

https://github.com/OneSignal/OneSignal-Unity-SDK/issues/123

**Comment by willispinaud** [comment]
Same problem here

**Comment by jkasten2** [comment]
Can you try importing `FirebaseMessaging.unitypackage` to your project from https://firebase.google.com/docs/unity/setup? I believe there is an dependency issue with `FirebaseAnalytics.unitypackage`.

**Comment by jkasten2** [comment]
@Saoish Good to hear that solved your issue. I have updated our docs to make a note of this.
https://documentation.onesignal.com/docs/unity-sdk-setup#section-3-android-setup

It seems you must import `FirebaseMessaging.unitypackage` if you use **any** Firebase Unity plugins.

@stewartmiles
I have tried reporting this compatibility issue to Firebase Support but they noted to me that it is out of their [support scope](https://firebase.google.com/support/scope#integrating_with_third_party_libraries). I know this isn't in the scope of `unity-jar-resolver`, `quickstart-unity`, or other projects you work on but if there is anything you can do from your end that would be great.
In summary the issue is the Firebase Unity SDK seems to throw an initialization error when the official `.firebase-messaging.aar` is present in the project, without any 3rd party code. Example added this .aar will break the Unity Firebase Analytics SDK in the app.
**Steps to reproduce with the quickstart app;**
1. Follow the [Unity Firebase Analytics quickstart setup](https://github.com/firebase/quickstart-unity/tree/master/analytics/testapp) guide and ensure your analytics calls are working.
2. Create a folder `Assets/PluginExample/Editor/`
3. Create a new file here name it `PluginExampleDependencies.xml`
4. With the following contents
```xml
<dependencies>
  <androidPackages>
    <androidPackage spec="com.google.firebase:firebase-messaging:17.0.0">
    </androidPackage>
  </androidPackages>
</dependencies>
```
5. Go to `Assets` -> `Play Services Resolver` -> `Android Resolver` -> `Force Resolve`
6. To ensure steps 2 through 5 were done correctly check that you have `Assets/Plugins/Android/com.google.firebase.firebase-messaging-17.3.0.aar`
7. Build and run your app again on the device.
8. Notice that Firebase analytics are no longer working and the errors in the logcat.
   * First thing you will notice is the "Show Analytics Instance ID" in the app no longer prints anything on the device.
9. The error in the logcat you will notice is the following
```
E/firebase: Java class com/google/firebase/messaging/cpp/RegistrationIntentService not found.
```

Thanks.

**Comment by stewartmiles** [comment]
@jkasten2 our code that auto-initializes parts of the C++ library enables Firebase Messaging initialization when the com/google/firebase/messaging/FirebaseMessaging is included in an application.  This is unfortunate as we also have a dependency upon the com.google.firebase.messaging.cpp package shipped with the C++ and Unity SDKs, when that is missing Firebase messaging initialization fails, throws an exception and stops everything from working.

We actually had an open issue to fix this, looks like this bumps the priority of fixing this now.  We're going to change the C++ module initialization to depend upon the presence of the included Firebase C# modules instead, that will allow the C++ component of messaging to lie dormant if the C# DLL isn't present in the project.

**Comment by jkasten2** [comment]
@stewartmiles Ah, was thinking that fixing the Firebase C++ library to look for presents of the C++ class `com.google.firebase.messaging.cpp` instead of the Java class `com/google/firebase/messaging/FirebaseMessaging` would fix the issue but C# check sounds like it will work too.

Thanks for reporting back and prioritizing! This will be a great fix to improve stability between SDKs!

**Comment by stewartmiles** [comment]
@jkasten2 we've fixed the issue in the 5.3.1 SDK release.
https://firebase.google.com/support/release-notes/unity

**Comment by jkasten2** [comment]
@stewartmiles Awesome! Thanks for the quick turn around!

We'll update our docs to recommend developers update to this version to ensure compatibility between Firebase and OneSignal.
